plugins/woocommerce/tests/php/src/Internal/Admin/ProductDownloadsPreviewTest.php

```diff
@@ -0,0 +1,237 @@
+		
+		$response = $this->preview->get_preview_permissions_check( $request );
+		
+		$this->assertInstanceOf( 'WP_Error', $response );
+		$this->assertEquals( 'woocommerce_rest_missing_token', $response->get_error_code() );
+	}
+
+	/**
+	 * Test permissions check fails with invalid token
+	 */
+	public function test_get_preview_permissions_check_fails_with_invalid_token() {
+		$request = new WP_REST_Request( 'GET', '/wc/v3/admin/product-downloads-preview/' . $this->product_id . '/' . $this->attachment_id );
+		$request->set_param( 'token', 'invalid_token' );
+		
+		$response = $this->preview->get_preview_permissions_check( $request );
+		
+		$this->assertInstanceOf( 'WP_Error', $response );
+		$this->assertEquals( 'woocommerce_rest_invalid_token', $response->get_error_code() );
+	}
+
+	/**
+	 * Test permissions check fails with mismatched resources
+	 */
+	public function test_get_preview_permissions_check_fails_with_mismatched_resources() {
+		// Generate a token
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		wp_set_current_user( $user_id );
+
+		// Create a token for different product
+		$different_product_id = $this->product_id + 1;
+		$token = wp_generate_password( 32, false );
+		set_transient(
+			'wc_preview_token_' . $token,
+			array(
+				'attachment_id' => $this->attachment_id,
+				'product_id'    => $different_product_id,
+			),
+			60
+		);
+		
+		$request = new WP_REST_Request( 'GET', '/wc/v3/admin/product-downloads-preview/' . $this->product_id . '/' . $this->attachment_id );
+		$request->set_param( 'token', $token );
+		$request->set_param( 'product_id', $this->product_id );
+		$request->set_param( 'attachment_id', $this->attachment_id );
+		
+		$response = $this->preview->get_preview_permissions_check( $request );
+		
+		$this->assertInstanceOf( 'WP_Error', $response );
+		$this->assertEquals( 'woocommerce_rest_token_mismatch', $response->get_error_code() );
+	}
+
+	/**
+	 * Test permissions check passes with valid token
+	 */
+	public function test_get_preview_permissions_check_passes_with_valid_token() {
+		// Generate a token
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		wp_set_current_user( $user_id );
+
+		$token = wp_generate_password( 32, false );
+		set_transient(
+			'wc_preview_token_' . $token,
+			array(
+				'attachment_id' => $this->attachment_id,
+				'product_id'    => $this->product_id,
+			),
+			60
+		);
+		
+		$request = new WP_REST_Request( 'GET', '/wc/v3/admin/product-downloads-preview/' . $this->product_id . '/' . $this->attachment_id );
+		$request->set_param( 'token', $token );
+		$request->set_param( 'product_id', $this->product_id );
+		$request->set_param( 'attachment_id', $this->attachment_id );
+		
+		$response = $this->preview->get_preview_permissions_check( $request );
+		
+		$this->assertTrue( $response );
+		
+		// Verify token was deleted after use
+		$this->assertFalse( get_transient( 'wc_preview_token_' . $token ) );
```

Comment:
Feel free to disregard ... this is a bit of a quibble. We're testing an implementation detail here (ie, the transient used to store the one-time token has been deleted). 

Could demonstrating that the same request won't succeed twice be better, or a worthwhile supplement to this test? In my mind, that describes the outcome a little more clearly.

Reply:
I agree with you, no need to test that the cache/transient was stored. I'll simplify the testing in a sect.

Reply:
Removed in f75bf08010

Reply:
I've also added the new tests to make sure we're not able to get the same URL more than once in 62a4aae87d .

Reply:
And then I removed it again as I've opted to use a simpler mechanism for access control, so no need to worry about access counts. Admin can access it as many times as they want.